### ISSUE TITLE
VS2017 Windows 10 Fails to build

### ISSUE THREAD

**Issue by soviettoly** [RobertBeckebans/RBDOOM-3-BFG#425]
I get the error below when trying to build either idlib or the game itself. For some reason the error comes to xcall_once.h which is in msvc lib it seems. 

> 2>C:\ProgramFiles (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.15.26726\include\xcall_once.h(20): error C3615: constexpr function 'std::once_flag::once_flag' cannot result in a constant expression
> 2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.15.26726\include\xcall_once.h(21): note: failure was caused by call of undefined function or one not declared 'constexpr'
> 2>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.15.26726\include\xcall_once.h(21): note: see usage of 'idNullPtr::operator T1'
> 2>        with
> 2>        [
> 2>            T1=void
> 2>        ]
> 2>Done building project "idlib.vcxproj" -- FAILED.

**Comment by valkarion** [comment]
Hello, I had the same issue, and the error message also hints at it. In sys_types.h there is an idNullPtr struct defined. it has a `template<typename T1> operator T1* () const` method defined. What you need to do is add a constexpr modifier to it, so it looks like this: `template<typename T1> constexpr operator T1* () const` after this my version successfully built.

**Comment by Phariax** [comment]
> Hello, I had the same issue, and the error message also hints at it. In sys_types.h there is an idNullPtr struct defined. it has a `template<typename T1> operator T1* () const` method defined. What you need to do is add a constexpr modifier to it, so it looks like this: `template<typename T1> constexpr operator T1* () const` after this my version successfully built.

Thank you valkarion, that was useful, it works.

**Comment by noabody** [comment]
```diff
diff --git a/neo/idlib/sys/sys_types.h b/neo/idlib/sys/sys_types.h
index e49551a..7751586 100644
--- a/neo/idlib/sys/sys_types.h
+++ b/neo/idlib/sys/sys_types.h
@@ -98,7 +98,7 @@ struct idNullPtr
 	idNullPtr() : value( 0 ) { }
 	
 	// implicit conversion to all pointer types
-	template<typename T1> operator T1* () const
+	template<typename T1> constexpr operator T1* () const
 	{
 		return 0;
 	}
```

**Comment by noabody** [comment]
See also this patch against the main project as it incorporates the fix noted plus a few others:
[Fix compilation in VS2017 #7](https://patch-diff.githubusercontent.com/raw/id-Software/DOOM-3-BFG/pull/7.patch)
The relevant portion appears to be:
```diff
diff --git a/neo/idlib/sys/sys_includes.h b/neo/idlib/sys/sys_includes.h
index 5008c82..56d4d89 100644
--- a/neo/idlib/sys/sys_includes.h
+++ b/neo/idlib/sys/sys_includes.h
@@ -48,8 +48,6 @@ If you have questions concerning this license or the applicable additional terms
 #ifndef _D3SDK
 #ifndef GAME_DLL
 
-#define WINVER				0x501
-
 #include <winsock2.h>
 #include <mmsystem.h>
 #include <mmreg.h>
diff --git a/neo/idlib/sys/sys_types.h b/neo/idlib/sys/sys_types.h
index e49551a..a740e02 100644
--- a/neo/idlib/sys/sys_types.h
+++ b/neo/idlib/sys/sys_types.h
@@ -95,16 +95,16 @@ struct idNullPtr
 {
 	// one pointer member initialized to zero so you can pass NULL as a vararg
 	void* value;
-	idNullPtr() : value( 0 ) { }
+	constexpr idNullPtr() : value( 0 ) { }
 	
 	// implicit conversion to all pointer types
-	template<typename T1> operator T1* () const
+	template<typename T1> constexpr operator T1 * () const
 	{
 		return 0;
 	}
 	
 	// implicit conversion to all pointer to member types
-	template<typename T1, typename T2> operator T1 T2::* () const
+	template<typename T1, typename T2> constexpr operator T1 T2::* () const
 	{
 		return 0;
 	}

```

**Comment by RobertBeckebans** [comment]
Those fixes are in the master branch.
